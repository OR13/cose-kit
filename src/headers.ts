import { fromUTF8 } from "./lib/buffer_utils.js";
import { encoder } from "./cbor.js";

export const algs = new Map<number, { name: string, hash?: string }>(
  [
    [-8, { name: 'EdDSA' }],
    [-7, { name: 'ES256', hash: 'SHA-256' }],
    [-35, { name: 'ES384', hash: 'SHA-384' }],
    [-36, { name: 'ES512', hash: 'SHA-512' }],
    [-37, { name: 'PS256', hash: 'SHA-256' }],
    [-38, { name: 'PS384', hash: 'SHA-384' }],
    [-39, { name: 'PS512', hash: 'SHA-512' }],
    [-257, { name: 'RS256', hash: 'SHA-256' }],
    [-258, { name: 'RS384', hash: 'SHA-384' }],
    [-259, { name: 'RS512', hash: 'SHA-512' }],
  ]
);

/**
   +-----------+-------+---------+----------+--------------------------+
   | Name      | Value | Hash    | Tag      | Description              |
   |           |       |         | Length   |                          |
   +-----------+-------+---------+----------+--------------------------+
   | HMAC      | 4     | SHA-256 | 64       | HMAC w/ SHA-256          |
   | 256/64    |       |         |          | truncated to 64 bits     |
   | HMAC      | 5     | SHA-256 | 256      | HMAC w/ SHA-256          |
   | 256/256   |       |         |          |                          |
   | HMAC      | 6     | SHA-384 | 384      | HMAC w/ SHA-384          |
   | 384/384   |       |         |          |                          |
   | HMAC      | 7     | SHA-512 | 512      | HMAC w/ SHA-512          |
   | 512/512   |       |         |          |                          |
   +-----------+-------+---------+----------+--------------------------+
 */
export const macAlgs = new Map<number, { name: string, hash: string, length?: number }>([
  [5, { name: 'HS256', hash: 'SHA-256', length: 256 }]
]);

export const macAlgsToValue = new Map<string, number>(
  [
    ['HS256', 5],
  ]
);

export const algsToValue = new Map<string, number>(
  [
    ['EdDSA', -8],
    ['ES256', -7],
    ['ES384', -35],
    ['ES512', -36],
    ['PS256', -37],
    ['PS384', -38],
    ['PS512', -39],
    ['RS256', -257],
    ['RS384', -258],
    ['RS512', -259],
  ]
);

export const headers: { [k: string]: number } = {
  partyUNonce: -22,
  static_key_id: -3,
  static_key: -2,
  ephemeral_key: -1,
  alg: 1,
  crit: 2,
  ctyp: 3,
  kid: 4,
  IV: 5,
  Partial_IV: 6,
  counter_signature: 7,
  x5bag: 32,
  x5chain: 33,
  x5t: 34,
  x5u: 35,
};

export type ProtectedHeader = {
  alg?: 'EdDSA' | 'ES256' | 'ES384' | 'ES512' | 'PS256' | 'PS384' | 'PS512' | 'RS256' | 'RS384' | 'RS512',
  crit?: number[],
  ctyp?: number | string,
  [key: string]: unknown,
};

export type MacProtectedHeader = {
  alg?: 'HS256',
  crit?: number[],
  ctyp?: number | string,
  [key: string]: unknown,
};

export type UnprotectedHeaders = {
  ctyp?: number | string,
  kid?: Uint8Array | string,
  x5chain?: Uint8Array | Uint8Array[],
};

export const encodeProtectedHeaders = (protectedHeader: ProtectedHeader | undefined) => {
  if (typeof protectedHeader === 'undefined') {
    return new Uint8Array();
  }

  return encoder.encode(new Map(Object.entries(protectedHeader || {}).map(([k, v]: [string, unknown]) => {
    if (k === 'alg') {
      v = algsToValue.get(v as string);
    } else if (typeof v === 'string') {
      v = fromUTF8(v);
    }
    return [headers[k], v];
  })))
};

export const mapUnprotectedHeaders = (unprotectedHeader: UnprotectedHeaders | undefined) => {
  return new Map(Object.entries(unprotectedHeader || {}).map(([k, v]: [string, unknown]) => {
    if (typeof v === 'string') {
      v = fromUTF8(v);
    }
    return [headers[k], v];
  }));
};
