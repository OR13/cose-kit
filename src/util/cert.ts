import { encodeBase64 } from '#runtime/base64.js';

export const pemToCert = (cert: string): string => {
  const pem = /-----BEGIN (\w*)-----([^-]*)-----END (\w*)-----/g.exec(cert.toString());
  if (pem && pem.length > 0) {
    return pem[2].replace(/[\n|\r\n]/g, '');
  }
  return '';
};

export const certToPEM = (cert: Uint8Array): string => {
  return `-----BEGIN CERTIFICATE-----
${encodeBase64(cert).trim()}
-----END CERTIFICATE-----`;
}
