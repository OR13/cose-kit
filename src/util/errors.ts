/**
 * A generic Error that all other COSE specific Error subclasses extend.
 *
 * @example Checking thrown error is a COSE one
 *
 * ```js
 * if (err instanceof cose.errors.COSEError) {
 *   // ...
 * }
 * ```
 */
export class COSEError extends Error {
  /** A unique error code for the particular error subclass. */
  static get code(): string {
    return 'ERR_COSE_GENERIC'
  }

  /** A unique error code for the particular error subclass. */
  code: string = 'ERR_COSE_GENERIC'

  constructor(message?: string) {
    super(message)
    this.name = this.constructor.name
    // @ts-ignore
    Error.captureStackTrace?.(this, this.constructor)
  }
}

/**
 * An error subclass thrown when a JOSE Algorithm is not allowed per developer preference.
 *
 * @example Checking thrown error is this one using a stable error code
 *
 * ```js
 * if (err.code === 'ERR_COSE_ALG_NOT_ALLOWED') {
 *   // ...
 * }
 * ```
 *
 * @example Checking thrown error is this one using `instanceof`
 *
 * ```js
 * if (err instanceof jose.errors.JOSEAlgNotAllowed) {
 *   // ...
 * }
 * ```
 */
export class COSEAlgNotAllowed extends COSEError {
  static get code(): 'ERR_COSE_ALG_NOT_ALLOWED' {
    return 'ERR_COSE_ALG_NOT_ALLOWED'
  }

  code = 'ERR_COSE_ALG_NOT_ALLOWED'
}

/**
 * An error subclass thrown when a particular feature or algorithm is not supported by this
 * implementation or JOSE in general.
 *
 * @example Checking thrown error is this one using a stable error code
 *
 * ```js
 * if (err.code === 'ERR_COSE_NOT_SUPPORTED') {
 *   // ...
 * }
 * ```
 *
 * @example Checking thrown error is this one using `instanceof`
 *
 * ```js
 * if (err instanceof jose.errors.JOSENotSupported) {
 *   // ...
 * }
 * ```
 */
export class COSENotSupported extends COSEError {
  static get code(): 'ERR_COSE_NOT_SUPPORTED' {
    return 'ERR_COSE_NOT_SUPPORTED'
  }

  code = 'ERR_COSE_NOT_SUPPORTED'
}
