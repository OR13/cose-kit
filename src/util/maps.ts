/**
 * Creates a new map with the keys and values of the given map swapped.
 *
 * @param map {Map<K, V>} the original map
 * @returns {Map<V, K>} the new map
 */
export const reverseMap = <K, V>(map: Map<K, V>): Map<V, K> => new Map(Array.from(map).map(([k, v]) => [v, k]));
