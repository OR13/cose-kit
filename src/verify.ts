import { COSEVerifyGetKey } from "./jwks/local";
import { encoder } from "./cbor";
import { Sign1 } from "./cose/Sign1";
import { KeyLike } from "jose";
import { Sign } from "./cose/Sign";

type VerifyResult = {
  isValid: boolean;
  decoded: Sign1;
  key?: KeyLike | Uint8Array;
};

type MultiSigVerifyResult = {
  isValid: boolean;
  decoded: Sign;
  key?: KeyLike | Uint8Array;
};

export const coseVerify = async (
  cose: Uint8Array,
  key: KeyLike | Uint8Array | COSEVerifyGetKey
): Promise<VerifyResult> => {
  const decoded = encoder.decode(cose);

  if (!(decoded instanceof Sign1)) {
    throw new Error('unknown COSE type');
  }

  const isValid = await decoded.verify(key);

  return { isValid, decoded };
};

export const coseVerifyMultiSignature = async (
  cose: Uint8Array,
  keys: KeyLike[] | Uint8Array[] | COSEVerifyGetKey
): Promise<MultiSigVerifyResult> => {
  const decoded = encoder.decode(cose);

  if (!(decoded instanceof Sign)) {
    throw new Error('unexpected COSE type');
  }

  const isValid = await decoded.verify(keys);

  return { isValid, decoded };
};
