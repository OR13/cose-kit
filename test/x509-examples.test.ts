import * as fs from 'fs';
import crypto from 'crypto';

import { coseVerifyX509 } from '../src/verify.js';
import { Sign, coseMultiSign, coseSign } from '../src/index.js';
import { JWK, importJWK } from 'jose';
import { parseJWK } from './util.js';
import { headers } from '../src/headers.js';

const caRoots = [
  fs.readFileSync(`${__dirname}/Examples/x509-examples/ca.crt`, 'utf8')
];

const examples = [
  `${__dirname}/Examples/x509-examples/signed-03.json`,

].map((examplePath) => JSON.parse(fs.readFileSync(examplePath, 'utf8')));

describe('x509-examples', () => {

  examples.forEach(example => {

    describe(example.title, () => {

      it('should verify the signature', async () => {
        const result = await coseVerifyX509(
          Buffer.from(example.output.cbor, 'hex'),
          caRoots
        );
        expect(result.isValid).toBeTruthy();
      });

      describe('when signing a certificate with x5chain', () => {
        let signatureVerificationResult: Awaited<ReturnType<typeof coseVerifyX509>>;

        beforeAll(async () => {
          let signed: Uint8Array;
          if (example.input.sign0) {
            signed = await coseSign(
              example.input.sign0.protected,
              example.input.sign0.unprotected,
              Buffer.from(example.input.plaintext, 'utf8'),
              await importJWK(parseJWK(example.input.sign0.key))
            );
          } else {
            const signers = await Promise.all(
              example.input.sign.signers.map(async (signer: { key: JWK; protected: unknown; unprotected: { x5chain?: string }; }) => {
                return {
                  key: await importJWK(parseJWK(signer.key)),
                  protectedHeaders: signer.protected,
                  unprotectedHeaders: {
                    ...signer.unprotected || {},
                    x5chain: signer.unprotected?.x5chain ? Buffer.from(signer.unprotected?.x5chain, 'hex') : undefined
                  },
                };
              }
              )
            );
            signed = await coseMultiSign(
              example.input.sign.protected,
              example.input.sign.unprotected,
              Buffer.from(example.input.plaintext, 'utf8'),
              signers
            );
          }
          signatureVerificationResult = await coseVerifyX509(signed, caRoots);
        });

        it('should verify the signature', async () => {
          expect(signatureVerificationResult.isValid).toBeTruthy();
        });

        it('should contain the x5chain', () => {
          const current = ((signatureVerificationResult.decoded as Sign)
            .signatures[0]
            .unprotectedHeaders
            .get(headers.x5chain) as Buffer).toString('hex').toUpperCase();
          const expected = example.input.sign0?.unprotected?.x5chain || example.input.sign?.signers[0].unprotected.x5chain;
          expect(current).toBe(expected);
        });

      });
    });

  });

  describe('sign with an expired cert in x5chain', () => {
    const key = crypto.createPrivateKey(fs.readFileSync(`${__dirname}/Examples/x509-examples/expired.key`, 'utf8'));
    const x5chain = fs.readFileSync(`${__dirname}/Examples/x509-examples/expired.der`);
    const payload = Buffer.from('Hello World!');
    it('should fail to verify', async () => {
      const signed = await coseSign(
        { alg: 'ES256' },
        { x5chain },
        payload,
        key
      );
      await expect(coseVerifyX509(signed, caRoots))
        .rejects
        .toThrowErrorMatchingSnapshot();
    });
  });
});
