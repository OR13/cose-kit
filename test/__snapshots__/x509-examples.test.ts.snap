// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`x509-examples sign with an expired cert in x5chain should fail to verify 1`] = `"The certificate is either not yet valid or expired"`;
